# Basic `pandas`

In this notebook we will learn about `pandas` a vital data management package in python.

By the end of this notebook you will know about:
- `pandas` `Series` and `DataFrame` objects,
- Indexing `pandas` objects,
- Reading files with `pandas`,
- Saving files with `pandas` and 
- Some helpful `pandas` functionality.

## `pandas`

`pandas` is one of the most popular data handling packages in `python`. We will cover the minimum you need to know about the package for the boot camp in this notebook.

Let's start by importing the package. Just like in the previous notebook this should check whether or not you have `pandas` installed on your machine. If you are using Anaconda, <a href="https://www.anaconda.com/">https://www.anaconda.com/</a>, `pandas` should be installed already. If not check out the documentation for installation instructions <a href="https://pandas.pydata.org/docs/getting_started/install.html">https://pandas.pydata.org/docs/getting_started/install.html</a>.

In [2]:
## it is standard to import pandas as pd

import pandas as pd


In [4]:
## let's check what version of numpy you have
## when I wrote this I had version 1.5.3
## yours may be different

print(pd)

<module 'pandas' from '/opt/anaconda3/lib/python3.12/site-packages/pandas/__init__.py'>


If you had a version of `pandas` installed, both of those code chunks should have executed without error. If not, you will need to install it onto your machine because we will be using it heavily in the boot camp. If you are unsure how to install a python package in general check our python package installation guide, <a href="https://www.erdosinstitute.org/data-science">https://www.erdosinstitute.org/data-science</a>.

##### Be sure you can run both of the above code chunks before continuing with this notebook, again it should be fine if your package version is slightly different than mine.

### `Series` and `DataFrame`s

`pandas` has two main data structures: 
1. `Series` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</a> and 
2. `DataFrame` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html</a>. 

Let's explore them below.

In [8]:
## We can turn a list into a series
## with pd.Series()
list1 = [0,1,2,3]
print("Below is a list.")
print(list1, type(list1))
print()
print("Below is a pandas Series object.")
print(pd.Series(list1))
type(pd.Series(list1))

Below is a list.
[0, 1, 2, 3] <class 'list'>

Below is a pandas Series object.
0    0
1    1
2    2
3    3
dtype: int64


pandas.core.series.Series

The second thing we printed was a `Series` object. Note the two columns of numbers. The first column is the index of the object, the second column contains the values of the object. We can access those two separately like below.

In [10]:
series1 = pd.Series(list1)

In [72]:
## The index

pd.Index(series1)

Index([0, 1, 2, 3], dtype='int64')

In [66]:
## The values
## I don't understand how to do this one. Something is missing. Look at documentation?

series1.values

array([0, 1, 2, 3])

In [18]:
## You practice
## Take the list labeled a and 
## turn it into a Series named b
a = [5,2,3,6,'a','b','e',True,False]

b = pd.Series(a)


In [20]:
## We can get select entries from a Series 
## using the index

pd.Index(b)

Index([5, 2, 3, 6, 'a', 'b', 'e', True, False], dtype='object')

In [28]:
## Note that we don't have to have our index be a list of numbers
c = pd.Series([1,2,3,4], index=['a','b','c','d'])

print(c)

a    1
b    2
c    3
d    4
dtype: int64


In [36]:
## and we can access by index in the same way

pd.Index(c)

Index([1, 2, 3, 4], dtype='int64')

Now let's check out a `DataFrame`. `DataFrame`s are essentially a collection of `Series` with a common index. We can also think of them as a table of values with column and row labels.

In [40]:
## We can make a DataFrame using a dictionary
## the dictionary keys are the column labels
## the dictionary values are columns
df = pd.DataFrame({'one':[3,4,5,2,4,5], 
                    'two':['a','b','e','h','l','p'],
                    'third column':[7,7,7,7,7,7]})

## NOTE that this is not the only way to make 
## a dataframe!

print(df)

   one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7


This is a `DataFrame`, the unlabeled column is the index, the labeled columns are `Series` objects themselves. We can access them in the following way.

In [74]:
## If your column's name doesn't violate a couple of format rules
## you can use
## df.column_name
print(df.one) 
print(df.two)
# cannot use print(df.third column) because of the space!
print()
print("Note that each column is a Series object.")
print(type(df.one))

0    3
1    4
2    5
3    2
4    4
5    5
Name: one, dtype: int64
0    a
1    b
2    e
3    h
4    l
5    p
Name: two, dtype: object

Note that each column is a Series object.
<class 'pandas.core.series.Series'>


In [ ]:
## However if our column name has spaces or certain characters in it
## (like . , ! ? "" and so on)
## we can't use .column_name


In [80]:
## So we have to use df["column name"] instead
# or df['column_name']

df["third column"]


0    7
1    7
2    7
3    7
4    7
5    7
Name: third column, dtype: int64

In [82]:
## Just like with a Series we can use .index
## to get the index
df.index

RangeIndex(start=0, stop=6, step=1)

In [76]:
## .values returns a 2-D numpy ndarray with our columns
df.values

array([[3, 'a', 7],
       [4, 'b', 7],
       [5, 'e', 7],
       [2, 'h', 7],
       [4, 'l', 7],
       [5, 'p', 7]], dtype=object)

In [102]:
## You code
## Make a data frame, call it my_df 
## Make the first column labeled 'first' from a
## Make the second column labeled 'second' from b
## see what happens when you add , index=range(10,10+len(a)) 
## after the dictionary

a = [4,5,3,4,5,6,0]
b = ['a','c','d','g','l','m','p']

my_df=pd.DataFrame({'first':a,'second':b}, index=range(10,10+len(a)))

my_df.values

array([[4, 'a'],
       [5, 'c'],
       [3, 'd'],
       [4, 'g'],
       [5, 'l'],
       [6, 'm'],
       [0, 'p']], dtype=object)

In [104]:
my_df

first second
10      4      a
11      5      c
12      3      d
13      4      g
14      5      l
15      6      m
16      0      p

#### Locating `DataFrame` Entries With the Index

Locating entries in a `DataFrame` is slightly more complicated than what we had to do for `Series` objects.

In [168]:
## You code
## try to index the dataframe df like you would
## a list or Series to find the entry corresponding
## to index 1
df[1]

## What happens?

# it dies.

KeyError: 1

In order to get particular `DataFrame` entries you have to use one of two methods, `.loc[]` or `.iloc[]`.

In [110]:
df.loc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

In [112]:
## for example we can get the 1 row of df like so with loc

df.loc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

In [116]:
## or like so with iloc

df.iloc[0]

one             3
two             a
third column    7
Name: 0, dtype: object

The difference between `.loc` and `.iloc` is that `.iloc` is restricted to integer based indexing, while `.loc` is more versatile.

This means that if your index is a list of strings you cannot use `.iloc` with just the index's normal value. For example try to run the following.

In [118]:
## Recall what c is

c

a    1
b    2
c    3
d    4
dtype: int64

In [146]:
## now try to get the ath entry using .iloc

c.iloc[a]

IndexError: positional indexers are out-of-bounds

In [142]:
## instead we'd have to notice that the "a" entry is
## the 0 entry if the index were a normal integer based index
c.iloc[0]

1

In [156]:
## However, we can just insert "a" into .loc
c.loc["a"]

1

Another difference is that `.loc` can be used to perform boolean indexing (eg. find me all the rows where the first column is between $2$ and $4$)

In [198]:
df


one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [200]:
df.values

array([[3, 'a', 7],
       [4, 'b', 7],
       [5, 'e', 7],
       [2, 'h', 7],
       [4, 'l', 7],
       [5, 'p', 7]], dtype=object)

In [216]:
## give me all the rows where the "one" column is between 2 and 4
df.loc[(df.one <= 4) & (df.one >= 2), ['one']]

# Works: df.loc[(df.one < 4) & (df.one > 2), ['third column','two']]
# iloc only handles integer indexing.

one
0    3
1    4
3    2
4    4

In [164]:
## we can even go one step further and find a specific column
df.loc[0]

one             3
two             a
third column    7
Name: 0, dtype: object

In [256]:
## or a subset of many columns
## in this case you enter the subset as a list of column name strings
df.loc[[0,3]]

df.loc[[0,1,2,3],['one','two']]

one two
0    3   a
1    4   b
2    5   e
3    2   h

In [206]:
## order of columns can even be different from dataframe
df.loc[(df.one < 4) & (df.one > 2), ['third column','two']]

third column two
0             7   a

### Reading in a `.csv` file with `pandas`

You can also read in common data file types with `pandas`. Let's load the following.

In [258]:
## You can read in a csv with
## pd.read_csv("filename.csv")
## https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

## We'll work with the following dataframe in the next section
jr_shots = pd.read_csv("../data/JR_Smith_Shots_2015_16.csv")

In [ ]:
jr_shots

In [262]:
## You code
## read in the file "beer.csv" using pandas
## call it beer_df

beer_df = pd.read_csv("../data/beer.csv")

## you'll practice with this dataframe in the next section

In [ ]:
beer_df

There are other useful `read` functions like `read_table` and `read_json`, you will see those later.

### Helpful `pandas` functions

`pandas` offers some really nice built in functions to help you explore any data set you're dealing with. Let's explore them below.

In [264]:
## df.head(n) let's you inspect the first n rows of the dataframe
## n defaults to 5
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html
jr_shots.head()

LOC_X  LOC_Y  SHOT_MADE_FLAG
0   -106    244               0
1    -96     97               0
2     30     23               0
3   -204     -1               0
4    -76    237               0

In [270]:
## You code 
## investigate what .tail(n) does 
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html
## Hint: n must be an integer

beer_df.tail(5)

IBU  ABV  Rating Beer_Type
342   68  8.2   3.922       IPA
343   45  6.8   4.000       IPA
344  101  9.2   3.778       IPA
345   70  6.8   4.015       IPA
346   80  6.9   3.554       IPA

In [282]:
## You code
## investigate what .sample(n) does
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html
## Hint: n must be an integer

beer_df.sample(2)


IBU  ABV  Rating Beer_Type
98   20  7.5   3.671     Stout
88   12  7.8   3.905     Stout

In [318]:
## You code
## what happens when you add the argument
## random_state=440 in .sample()?
## rerun the code chunk multiple times with or without the random_state
## try different numbers for the random_state

beer_df.sample(2,random_state=20)

# random_state fixes the random state basically. You get the same information back each time.

IBU   ABV  Rating Beer_Type
63    60   9.7   3.856     Stout
119   55  11.0   4.075     Stout

In [324]:
## df.info() tells you useful information about your
## dataframe, like the column names,
## the number of rows,
## the number of non-empty entries for each column,
## the data type of each column
## and how much memory the dataframe uses
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html
jr_shots.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 849 entries, 0 to 848
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   LOC_X           849 non-null    int64
 1   LOC_Y           849 non-null    int64
 2   SHOT_MADE_FLAG  849 non-null    int64
dtypes: int64(3)
memory usage: 20.0 KB


In [326]:
## df.describe()
## provides the 5 stat breakdown of each numeric column
## that is any column that consists of integers or floats
## will have its, minimum, maximum, first quartile, median,
## third quartile, and mean provided
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
jr_shots.describe()

# Gives useful statistical information!!

LOC_X       LOC_Y  SHOT_MADE_FLAG
count  849.000000  849.000000      849.000000
mean    -6.204947  112.106007        0.414605
std    157.201943   91.254429        0.492944
min   -250.000000  -46.000000        0.000000
25%   -158.000000   21.000000        0.000000
50%     -2.000000  110.000000        0.000000
75%    133.000000  193.000000        1.000000
max    245.000000  693.000000        1.000000

In [358]:
## You code
## Try out .mean() and .max() and .min()
## on beer_df
## Compare the results to .describe()'s output

## .mean() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
## .max() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html
## .min() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.min.html

#beer_df.mean()

# beef_df.mean() doesn't work?

# I guess it tries to give the numerical information from all columns but it canot do it for the type of beer.

beer_df.describe()

IBU         ABV      Rating
count  347.000000  347.000000  347.000000
mean    57.579251    7.639769    3.853980
std     24.100858    2.216800    0.228915
min      6.000000    3.800000    3.022000
25%     40.000000    6.200000    3.700500
50%     60.000000    7.000000    3.823000
75%     70.000000    8.850000    3.982500
max    250.000000   18.000000    4.541000

In [360]:
## What about variables that aren't numeric? Like
## classes
## You can use df.value_counts() for those
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html
jr_shots.SHOT_MADE_FLAG.value_counts()

#it gives the values of shot made and not.


SHOT_MADE_FLAG
0    497
1    352
Name: count, dtype: int64

The above tells us that 497 observations have a `SHOT_MADE_FLAG` of 0 and 352 have a `SHOT_MADE_FLAG` of 1.

In [372]:
## You code
## How many of each "Beer_Type" are there in beer_df?

beer_df.Beer_Type.value_counts(normalize=True)

beer_df.Beer_Type.value_counts()

## What happens when you input the argument normalize=True into value_counts()?

#normalize=True counts the proportion of eacht ype.


Beer_Type
IPA      195
Stout    152
Name: count, dtype: int64

One task that we will do quite often in the boot camp is split one data set into two non-overlapping data sets. Let's see how we can do that here.

In [438]:
## Let's first randomly sample 100 observations from jr_shots
## Note the .copy()
## this creates a deep copy of the DataFrame 
jr_1 = jr_shots.sample(100)

In [442]:
## To create jr_2 we'll just drop the indices of jr_1 from
## jr_shots using .drop()
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html

## Again we include .copy() to make a deep copy of the dataframe
jr_2 = jr_shots.drop(jr_1.index)

In [ ]:
jr_1

In [ ]:
jr_2

In [446]:
## Let's check that it worked like we thought it would
len(jr_1) + len(jr_2) == len(jr_shots)

True

In [448]:
## You code
## split beer_df into beer_1 and beer_2
## beer_1 should be a random sample of 150 observations
## beer_2 should be all rows not in beer_1


beer_1 = beer_df.sample(150)

beer_2 = beer_df.drop(beer_1.index)

In [458]:
len(beer_1)+len(beer_2) == len(beer_df)

True

### Saving a `DataFrame` to File

Just like we easily read in a data file we can use `pandas` to quickly save data to a file.

This is done with the function `df.to_csv()`, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html</a>.

In [460]:
## Let's first make a data frame
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [464]:
## Now we call df.to_csv(file_name)
df.to_csv("our_first_dataframe.csv")

#You must include the .csv of course.

Go ahead and check your repository, you should now see `our_first_dataframe.csv` in there.

In [482]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df

pd.read_csv("our_first_dataframe.csv")

## is anything off?

## Sure what is unnamed doing there.

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

When you ran `to_csv()` the default is to record the index as well. That is why your `DataFrame` has an unlabeled column when you read it back in using `read_csv`. If we wanted to avoid writing the index to file you include the argument `index=False`. 

In [468]:
## Make the DataFrame again
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [478]:
## write it to file without the index
df.to_csv("our_first_dataframe.csv",index=False)

In [480]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df again

pd.read_csv("our_first_dataframe.csv")

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

That's all you will need to know from `pandas` to have a firm footing for our boot camp. We will learn more stuff as we go along, but when that time comes you should be ready!

--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2023.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)